source/sundials/arkode.cc

```diff
@@ -499,20 +1011,138 @@ namespace SUNDIALS
     };
   }
 
+  template <typename VectorType>
+  N_Vector
+  ARKode<VectorType>::create_vector(const VectorType &template_vector) const
+  {
+    N_Vector new_vector;
+    auto     system_size = template_vector.size();
+
+#  ifdef DEAL_II_WITH_MPI
+    if (is_serial_vector<VectorType>::value == false)
+      {
+        const IndexSet    is = template_vector.locally_owned_elements();
+        const std::size_t local_system_size = is.n_elements();
+
+        new_vector =
+          N_VNew_Parallel(communicator, local_system_size, system_size);
+      }
+    else
+#  endif
+      {
+        new_vector = N_VNew_Serial(system_size);
+      }
+    return new_vector;
+  }
+
+  template <typename VectorType>
+  void *
+  ARKode<VectorType>::get_arkode_memory() const
+  {
+    return arkode_mem;
+  }
+
+#  if DEAL_II_SUNDIALS_VERSION_GTE(5, 4, 0)
+  template <typename VectorType>
+  SundialsOperator<VectorType>::SundialsOperator(ARKode<VectorType> &solver,
+                                                 void *              A_data,
+                                                 ATimesFn            a_times_fn)
+    : solver(solver)
+    , A_data(A_data)
+    , a_times_fn(a_times_fn)
+
+  {
+    AssertThrow(a_times_fn != nullptr, ExcInternalError());
+  }
+
+  template <typename VectorType>
+  void
+  SundialsOperator<VectorType>::vmult(VectorType &      dst,
+                                      const VectorType &src) const
+  {
+    N_Vector sun_dst = solver.create_vector(dst);
+    N_Vector sun_src = solver.create_vector(src);
+    copy(sun_src, src);
+    int status = a_times_fn(A_data, sun_src, sun_dst);
+    (void)status;
+    AssertARKode(status);
+
+    copy(dst, sun_dst);
+  }
+
+  template <typename VectorType>
+  SundialsPreconditioner<VectorType>::SundialsPreconditioner(
+    ARKode<VectorType> &solver,
+    void *              P_data,
+    PSolveFn            p_solve_fn,
+    double              tol)
+    : solver(solver)
+    , P_data(P_data)
+    , p_solve_fn(p_solve_fn)
+    , tol(tol)
+  {}
+
+  template <typename VectorType>
+  void
+  SundialsPreconditioner<VectorType>::vmult(VectorType &      dst,
+                                            const VectorType &src) const
+  {
+    // apply identity preconditioner if nothing else specified
+    if (!p_solve_fn)
+      {
+        dst = src;
+        return;
+      }
+
+    N_Vector sun_dst = solver.create_vector(dst);
+    N_Vector sun_src = solver.create_vector(src);
+    copy(sun_src, src);
+    // TODO currently no distinction between left and right preconditioning is
+    // possible
+    int status = p_solve_fn(P_data, sun_src, sun_dst, tol, 0);
+    (void)status;
+    AssertARKode(status);
+
+    copy(dst, sun_dst);
+  }
+#  endif
+
   template class ARKode<Vector<double>>;
   template class ARKode<BlockVector<double>>;
 
+#  if DEAL_II_SUNDIALS_VERSION_GTE(5, 4, 0)
+  template struct SundialsOperator<Vector<double>>;
+  template struct SundialsOperator<BlockVector<double>>;
```

Comment:
They weren't instantiated before so I didn't add them for the new classes. Should I?

Reply:
Have you tried out to use `L:d:V`? Did it compile?

Reply:
It will not work out of the box because the `copy ` functions are not overloaded for L:d:V. Since I would like to get rid of copies, I guess one would enable it only after the N_Vector is introduced?

Reply:
Yes.